### ISSUE TITLE
Smartparents is getting super-slow

### ISSUE THREAD

**Issue by mrkkrp** [Fuco1/smartparens#571]
I've noticed that my Emacs is getting quite slow, so slow that when I type fast it lags sometimes noticeably. With help of built-in profiler, I've found out that if I press one letter (normally it would result in real-time update of display, but it lags instead), I get this distribution of cycles:

```
- command-execute                                               16575  67%
 - call-interactively                                           16570  67%
  - funcall-interactively                                       16570  67%
   - self-insert-command                                        15320  62%
    - sp--post-self-insert-hook-handler                         14793  60%
     + sp--all-pairs-to-insert                                   8478  34%
     + sp-insert-pair                                            6254  25%
     + sp-skip-closing-pair                                        30   0%
    + jit-lock-after-change                                       315   1%
    + flycheck-handle-change                                      182   0%
   + counsel-M-x                                                  893   3%
- delete-selection-pre-hook                                      5753  23%
 - apply                                                         5753  23%
  - ad-Advice-delete-selection-pre-hook                          5751  23%
   - sp-wrap--can-wrap-p                                         5751  23%
    - sp--get-pair-list-wrap                                     5711  23%
     - sp--do-action-p                                           5688  23%
      + sp-point-in-string                                       2776  11%
      + sp-point-in-comment                                      2731  11%
      + sp-get-pair                                                75   0%
      + -flatten                                                   26   0%
      sp--single-key-description                                   12   0%
+ flyspell-post-command-hook                                      964   3%
+ redisplay_internal (C function)                                 618   2%
+ sp--save-pre-command-state                                      326   1%
+ ...                                                             179   0%
+ yas--post-command-handler                                        16   0%
+ flycheck-hide-error-buffer                                       10   0%
+ flycheck-handle-signal                                            7   0%
+ flycheck-display-error-at-point-soon                              4   0%
  sml/generate-position-help                                        4   0%
+ timer-event-handler                                               4   0%
  internal-timer-start-idle                                         3   0%
  tooltip-hide                                                      3   0%
```

Note that 60% of time takes `sp--post-self-insert-hook-handler` and 23% takes `sp-wrap--can-wrap-p`. 83% of cycles are spent in Smartparens machinery.

----

I've experimentally confirmed that without the `smartparens` package, no slowdown occurs, with the following profiling results:

```
- flyspell-post-command-hook                                     3223  54%
 + flyspell-check-word-p                                         3036  51%
 + flyspell-check-pre-word-p                                        3   0%
   flyspell-word                                                    1   0%
+ redisplay_internal (C function)                                1852  31%
+ command-execute                                                 714  12%
+ ...                                                             103   1%
+ flycheck-display-error-at-point-soon                              7   0%
  flycheck-hide-error-buffer                                        4   0%
+ undo-auto--add-boundary                                           2   0%
```

Now `flyspell-post-command-hook` takes 54% of cycles, while previously its share was 3% because of Smartparens (assuming more or less constant speed of `flycheck-post-command-hook` we can talk about ×18 difference). Editing feels fast and pleasant, no lags.

----

This slow-down is not acceptable. If you can fix it, it would be great. Otherwise I fear I will have to replace Smartparens with something more efficient, even if not so functional.

**Comment by Fuco1** [comment]
I'm afraid I can't really asnwer much.  This doesn't happen to me.  Can you reproduce this in some minimal setup?

**Comment by mrkkrp** [comment]
This does not manifest itself right away, but after some editing the package seems to get into some state that is characterized by low performance. I will try to find a way to reproduce it reliably once I have time.

**Comment by mrkkrp** [comment]
OK, after some investigation it turns out that `gc-cons-threshold` has something to do with the statistic, although I have difficult time imagining that with value of 10240000 (default is 800000) holding a key for about 1 second can trigger GC (this turns out to be very reproducible with my setup) and thus freezes display update. With default settings, profiling shows that `self-insert-command` takes about 33%, display update takes 33%, and Flyspell gets the rest.

The thing is still slow, although display is updating continuously now. Smartparens does a lot after every key press adding to delay between physical interaction with keyboard and character appearing on the screen, which [may contribute to overall experience with editor in unconscious way](https://pavelfatin.com/typing-with-pleasure/). IMHO, it's not just Smartparens fault, but consequence of too many packages putting quite heavy stuff in `post-command-hook` and `post-self-insert-hook`; for example Flyspell is also quite slow.

The thing is frustrating, I recently tried Atom text editor, which kind of includes functionality of Smartparens by default. I got impression that although Emacs is coded in C and Elisp and should be faster than web-based Atom, when comparing them with all useful features enabled, Atom does feel more responsive when you just type text in it.

This may also have to do with Haskell mode. I have suspicion that it somehow adds to slowdown, because it's not that noticeable in other major modes. The last time I looked into code of Haskell mode it was quite low-quality in some places (it uses quite a few non-prefixed symbols, and in general feels like spaghetti that hasn't seen refactoring for years), and although there is some activity related to the mode recently, I believe it has no human resources to eliminate its technical debt and it seems like real pain to try to find out what might slow down things there.

Although frustrated, I have no complaints to Smartparens now. Closing.
